File: tlo_make_valid_filename.rb
#!/usr/bin/env ruby

INVALID_CHARACTERS = '\\/:*?"<>|'

def make_valid_filename(line, extension)
  line = line.dup
  extension = extension.dup

  line.strip!
  extension.strip!
  line << extension

  INVALID_CHARACTERS.each_char do |c|
    line.gsub!(c, "_")
  end

  if line.empty?
    raise Exception.new("error: empty filename")
  end

  return line
end

extension = ""
if ARGV.length > 0
  extension = ARGV[0].dup
end

begin
  STDIN.each do |line|
    line = make_valid_filename(line, extension)
    puts line
  end
rescue Exception => e
  puts e.message
  exit(1)
end

File: tlo_make_valid_filename.py
#!/usr/bin/env python3

import sys

INVALID_CHARACTERS = "\\/:*?\"<>|"

def make_valid_filename(line, extension):
    line = line.strip()
    extension = extension.strip()
    line += extension

    for c in INVALID_CHARACTERS:
        line = line.replace(c, "_")

    if not line:
        raise Exception("error: empty filename")

    return line

extension = ""
if len(sys.argv) > 1:
    extension = sys.argv[1]

try:
    for line in sys.stdin:
        line = make_valid_filename(line, extension)
        print(line)
except Exception as e:
    print(e)
    sys.exit(1)
